src/map/map.c

```diff
@@ -537,10 +537,18 @@ static int bl_vforeach(int (*func)(struct block_list*, va_list), int blockcount,
 	map->freeblock_lock();
 	for (i = blockcount; i < map->bl_list_count && returnCount < max; i++) {
 		if (map->bl_list[i]->prev) { //func() may delete this bl_list[] slot, checking for prev ensures it wasn't queued for deletion.
+			int returnVal = 0;
 			va_list argscopy;
+
 			va_copy(argscopy, args);
-			returnCount += func(map->bl_list[i], argscopy);
+			returnVal = func(map->bl_list[i], argscopy);
 			va_end(argscopy);
+
+			if (max == INT_MAX && returnVal == -1) {
```

Comment:
Why does this check specifically for `INT_MAX` (and is the `max` check necessary)?

Couldn't this just check for the return value (and document that in general, we allow to break out of a foreach loop by returning -1)

Reply:
[Note: as pointed out in irc, the `INT_MAX` check was necessary so that this will only affect the foreach functions and not the forcount functions, where a max count limit can be specified]